rust/ql/lib/codeql/rust/internal/Type.qll

```diff
@@ -624,3 +625,27 @@ final class ImplTraitTypeReprAbstraction extends TypeAbstraction, ImplTraitTypeR
     implTraitTypeParam(this, _, result.(TypeParamTypeParameter).getTypeParam())
   }
 }
+
+/**
+ * Holds if `root` is a valid complex [`self` root type][1], with type
+ * parameter `tp`.
+ *
+ * [1]: https://doc.rust-lang.org/stable/reference/items/associated-items.html#r-items.associated.fn.method.self-ty
+ */
+pragma[nomagic]
+predicate complexSelfRoot(Type root, TypeParameter tp) {
```

Comment:
What about removing the `tp` type parameter? That seems orthogonal to the purpose of picking the types that can appear for `self` and users can do `getPositionalTypeParameter(0)`?